mongo/integration/sessions_test.go

```diff
@@ -358,6 +360,115 @@ func TestSessions(t *testing.T) {
+			func(ctx context.Context) (string, error) {
+				_, err := mt.Coll.UpdateOne(ctx, bson.D{}, bson.D{{"$set", bson.D{{"a", 1}}}})
+				return "update", err
+			},
+			func(ctx context.Context) (string, error) {
+				model := mongo.NewUpdateOneModel().
+					SetFilter(bson.D{}).
+					SetUpdate(bson.D{{"$set", bson.D{{"a", 1}}}})
+				_, err := mt.Coll.BulkWrite(ctx, []mongo.WriteModel{model})
+				return "bulkWrite", err
+			},
+			func(ctx context.Context) (string, error) {
+				result := mt.Coll.FindOneAndDelete(ctx, bson.D{})
+				var err error
+				if err := result.Err(); err != nil && err != mongo.ErrNoDocuments {
+					err = err
+				}
+				return "findOneAndDelete", err
+			},
+			func(ctx context.Context) (string, error) {
+				result := mt.Coll.FindOneAndUpdate(ctx, bson.D{},
+					bson.D{{"$set", bson.D{{"a", 1}}}})
+				var err error
+				if err := result.Err(); err != nil && err != mongo.ErrNoDocuments {
+					err = err
+				}
+				return "findOneAndUpdate", err
+			},
+			func(ctx context.Context) (string, error) {
+				result := mt.Coll.FindOneAndReplace(ctx, bson.D{}, bson.D{{"a", 1}})
+				var err error
+				if err := result.Err(); err != nil && err != mongo.ErrNoDocuments {
+					err = err
+				}
+				return "findOneAndReplace", err
+			},
+			func(ctx context.Context) (string, error) {
+				operationName := "find"
+				cursor, err := mt.Coll.Find(ctx, bson.D{})
+				if err != nil {
+					return operationName, err
+				}
+				arr := bson.A{}
+				err = cursor.All(ctx, &arr)
+				return operationName, err
+			},
+		}
+
+		// maintainedOneSession asserts that exactly one session is used for all operations at least once
+		// across the retries of this test.
+		var maintainedOneSession bool
+
+		// limitedSessionUse asserts that the number of allocated sessions is strictly less than the number of
+		// concurrent operations in every retry of this test. In this instance it would be less than (but NOT
+		// equal to).
+		var limitedSessionUse bool
+
+		retrycount := 5
+		for i := 1; i <= retrycount; i++ {
+			errs, ctx := errgroup.WithContext(context.Background())
+
+			// Execute the ops list concurrently.
+			for _, op := range ops {
+				op := op
+				errs.Go(func() error {
+					cmd, err := op(ctx)
+					if err != nil {
+						return fmt.Errorf("error running %s operation: %v", cmd, err)
+					}
+					return nil
+				})
+			}
+			err := errs.Wait()
+			assert.Nil(mt, err, "expected no error, go: %v", err)
+
+			// Get all started events and collect them by the session ID.
+			set := make(map[string]bool)
+			for _, event := range mt.GetAllStartedEvents() {
+				lsid := event.Command.Lookup("lsid")
+				set[lsid.String()] = true
+			}
+			if len(set) == 1 {
+				maintainedOneSession = true
+			}
+			if len(set) < len(ops) {
+				limitedSessionUse = true
+			}
+		}
+
+		oneSessMsg := "expected one session accross all %v operations for at least 1/%v retries"
+		assert.True(mt, maintainedOneSession, oneSessMsg, len(ops), retrycount)
```

Comment:
@benjirewis @matthewdale @Neilshweky This is an important thing to note:

This assertion _always_ fails on evergreen for sharded clusters, and only sharded clusters. I can get it to create locally, but only if I add delays in the `(*session.Client).EndSession` method (`time.Sleep(3 *time.Second)`, for instance).  [This is a `MUST` requirement of the spec](https://github.com/mongodb/specifications/blob/master/source/sessions/tests/README.rst#implicit-sessions-only-allocate-their-server-session-after-a-successful-connection-checkout:~:text=Drivers%20MUST%20assert%20that%20exactly%20one%20session%20is%20used%20for%20all%20operations%20at%20least%20once%20across%20the%20retries%20of%20this%20test.), but the way the sequence is currently designed is not thread safe -- a race condition can occur when you end a session after ending a connection.

TL;DR: race conditions occur by virtue of the sequence of events.

How do we want to approach this? I think if we don't want to relax the conditions, we may have to create some kind of semaphore/mutex for ending sessions.

Reply:
Add DRIVERS-2398 , which seems to fix the problem for our tests. Thank you @matthewdale !